### DISCUSSION TITLE
Communication with via the on-board LED is sometimes confusing.

### DISCUSSION THREAD

**Discussion by factus10** [timex-sinclair-projects/TS-Pico#9] [Ideas]
Threads is experimental on MicroPython and can't be relied on to maintain LED status independently of the main process.

**Comment by Sasounet** [comment]
I know there has been discussion on this on the mailing list. It was said that the PICO can either work on something or blink the LED. 

We have the heart beat LED flashing slowly. It is reassuring, we know the system is working fine. During an operation, if we decide to have the LED being off, we don’t know if the PICO got stuck while the current heartbeat was off or if it is carrying the operation we asked it to do. Same thing if we decide to have the LED on during operation, we don’t know if it is stuck with the LED on or working.

In an ideal world here is what I would like best:

Slow flashing: heartbeat, the PICO is waiting for an instruction to process
Fast flashing: PICO is processing an instruction
Light stuck ON: PICO is stuck - there is a problem
Light stuck OFF: PICO is stuck - there is a problem.

This way there can’t be any confusion.

Now I know flashing the LED while doing something else is difficult, if not impossible. Can it be done through interrupts ? Any other way ?

François

**Comment by factus10** [comment]
François,

Yes, interrupts could work... threading in Micropython is a potential solution but it also has problems. Here's an example of using interrupts and both cores:
https://www.instructables.com/Dual-Cores-Interrupts-on-Pi-Pico/

The part that jumps out for me is this:
Always finish Core 1 before Core 0. If you interrupt with the STOP icon in Thonny this may only halt core 0 and leave core 1 running. This will cause a problem when you try to run the script again. 

Some of the LED blinking is handled by a thread on Core 1. This may be a part of some current bugs in the Pico.

Another approach is asynchio:
https://github.com/peterhinch/micropython-async/blob/master/v3/docs/TUTORIAL.md

This would require a bunch of code refactoring.

**Comment by rcalcagno** [comment]
Actually different LED blinking patterns are not so difficult to implement. If you look at the code, it's only a matter of changing the flash pattern interval (passed as a parameter to the BLINK_LED() routine) and call this in the second core whenever it's idle (most of the time). I am afraid I will disagree with David regarding second core occasionally interfering with the normal operation, and causing the Pico to take a nap; and would look at more mundane reasons, i.e. bad programming for instance.

Only thing should be modified is that as now, currently three fast blinks indicate an error during operation; we sould modify this somehow to indicate another condition.

Interrupts and asynchio are far overkilling IMHO, and in any case, as a friend of mine likes to put it, it's most likely a feature left for the students as a home assignment.

Sincerely,

Ricardo.
